src/Mvc/Mvc.RazorPages/src/Builder/RazorPagesEndpointRouteBuilderExtensions.cs

```diff
@@ -31,6 +32,245 @@ public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilde
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with a very low priority. The request will be routed to a page endpoint that
+        /// matches <paramref name="page"/>, and <paramref name="area"/>.
+        /// </summary>
+        /// <param name="routes">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="pattern">The route pattern.</param>
+        /// <param name="page">The action name.</param>
+        /// <param name="area">The area name.</param>
+        /// <remarks>
+        /// <para>
+        /// <see cref="MapFallbackToAreaPage(IEndpointRouteBuilder, string, string, string)"/> is intended to handle cases where URL path of
+        /// the request does not contain a filename, and no other endpoint has matched. This is convenient for routing
+        /// requests for dynamic content to a SPA framework, while also allowing requests for non-existent files to
+        /// result in an HTTP 404.
+        /// </para>
+        /// <para>
+        /// The order of the registered endpoint will be <c>int.MaxValue</c>.
+        /// </para>
+        /// <para>
+        /// This overload will use the provided <paramref name="pattern"/> verbatim. Use the <c>:nonfile</c> route contraint
+        /// to exclude requests for static files.
+        /// </para>
+        /// <para>
+        /// <see cref="MapFallbackToAreaPage(IEndpointRouteBuilder, string, string, string)"/> does not re-execute routing, and will
+        /// not generate route values based on routes defined elsewhere. When using this overload, the route values provided by matching 
+        /// <paramref name="pattern"/> will be available.
+        /// </para>
+        /// </remarks>
+        public static void MapFallbackToAreaPage(
+            this IEndpointRouteBuilder routes,
+            string pattern,
+            string page,
+            string area)
+        {
+            if (routes == null)
+            {
+                throw new ArgumentNullException(nameof(routes));
+            }
+
+            if (pattern == null)
+            {
+                throw new ArgumentNullException(nameof(pattern));
+            }
+
+            if (page == null)
+            {
+                throw new ArgumentNullException(nameof(page));
+            }
+
+            EnsureRazorPagesServices(routes);
+
+            // Called for side-effect to make sure that the data source is registered.
+            GetOrCreateDataSource(routes);
+
+            // Maps a fallback endpoint with an empty delegate. This is OK because
+            // we don't expect the delegate to run. 
+            routes.MapFallback(context => Task.CompletedTask).Add(b =>
+            {
+                // MVC registers a policy that looks for this metadata.
+                b.Metadata.Add(CreateDynamicPageMetadata(page, area));
+            });
+        }
```

Comment:
I'm really please how much powerful stuff can be built here with the building blocks we have available (ordering, constraints, policies, endpoints). None of these features required new features, it's all existing extensibility. @JamesNK - you're the best around.